clients/src/main/java/org/apache/kafka/common/record/KafkaBufferedInputStream.java

```diff
@@ -0,0 +1,484 @@
+     * {@code mark} method before subsequent calls to the
+     * {@code reset} method fail.
+     * Whenever the difference between {@code pos}
+     * and {@code markpos} exceeds {@code marklimit},
+     * then the  mark may be dropped by setting
+     * {@code markpos} to {@code -1}.
+     *
+     * @see     KafkaBufferedInputStream#mark(int)
+     * @see     KafkaBufferedInputStream#reset()
+     */
+    protected int marklimit;
+
+    /**
+     * Check to make sure that underlying input stream has not been
+     * nulled out due to close; if not return it;
+     */
+    private InputStream getInIfOpen() throws IOException {
+        InputStream input = in;
+        if (input == null)
+            throw new IOException("Stream closed");
+        return input;
+    }
+
+    /**
+     * Check to make sure that buffer has not been nulled out due to
+     * close; if not return it;
+     */
+    private byte[] getBufIfOpen() throws IOException {
+        byte[] buffer = buf;
+        if (buffer == null)
+            throw new IOException("Stream closed");
+        return buffer;
+    }
+
+    /**
+     * Creates a {@code BufferedInputStream}
+     * and saves its  argument, the input stream
+     * {@code in}, for later use. An internal
+     * buffer array is created and  stored in {@code buf}.
+     *
+     * @param   in   the underlying input stream.
+     */
+    public KafkaBufferedInputStream(InputStream in, BufferSupplier bufferSupplier) {
+        this(in, bufferSupplier, DEFAULT_BUFFER_SIZE);
+    }
+
+    /**
+     * Creates a {@code BufferedInputStream}
+     * with the specified buffer size,
+     * and saves its  argument, the input stream
+     * {@code in}, for later use.  An internal
+     * buffer array of length  {@code size}
+     * is created and stored in {@code buf}.
+     *
+     * @param   in     the underlying input stream.
+     * @param   size   the buffer size.
+     * @throws  IllegalArgumentException if {@code size <= 0}.
+     */
+    public KafkaBufferedInputStream(InputStream in, BufferSupplier bufferSupplier, int size) {
+        super(in);
+        if (size <= 0) {
+            throw new IllegalArgumentException("Buffer size <= 0");
+        }
+        this.bufferSupplier = bufferSupplier;
+        this.buffer = this.bufferSupplier.get(size);
+        this.buf = this.buffer.array(); //implicitly means off-heap buffers are not supported.
+    }
+
+    /**
+     * Fills the buffer with more data, taking into account
+     * shuffling and other tricks for dealing with marks.
+     * Assumes that it is being called by a synchronized method.
+     * This method also assumes that all data has already been read in,
+     * hence pos > count.
+     */
+    private void fill() throws IOException {
+        byte[] buffer = getBufIfOpen();
+        if (markpos < 0)
+            pos = 0;            /* no mark: throw away the buffer */
+        else if (pos >= buffer.length) { /* no room left in buffer */
```

Comment:
since markpos should always be <0, and thus the whole else if block after it should never be true, I prefer nuking the entire else-if block and add a comment explaining why it can be removed.